### PR TITLE
Private/devidask/attach node

### PR DESCRIPTION
**Pull request by devidask27**
added attach-node command

### CODE REVIEW

File: cmd/attachNode.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+var (
+	masterIPs   []string
+	workerIPs   []string
+	clusterName string
+)
+
+var attachNodeCmd = &cobra.Command{
+	Use:   "attach-node",
+	Short: "attaches node to kubernetes cluster",
+	Long:  "attaches node to kubernetes cluster",
```
Comment (anupbarve) [inline cmd/attachNode.go:24]: Need to add more information here.
Example: 
- About considerations around combinations of -m and -w working together

File: cmd/attachNode.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	Args: func(configCmdSet *cobra.Command, args []string) error {
+		if len(args) > 1 {
+			return errors.New("only one argument is required")
```
Comment (anupbarve) [inline cmd/attachNode.go:27]: Refine this to :
"Only cluster name is accepted as a parameter"

File: cmd/attachNode.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		} else if len(args) < 1 {
+			return errors.New("cluster name is required")
```
Comment (anupbarve) [inline cmd/attachNode.go:29]: Refine this to:
"Cluster name is required for attach-node"

File: cmd/attachNode.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+func hostId(exec cmdexec.Executor, fqdn string, token string, IPs []string) ([]string, error) {
+	var hostIdsList []string
+	tkn := fmt.Sprintf(`"X-Auth-Token: %v"`, token)
+	for _, ip := range IPs {
+		ip = fmt.Sprintf(`"%v"`, ip)
+		cmd := fmt.Sprintf("curl -sH %v -X GET %v/resmgr/v1/hosts | jq -r '.[] | select(.info.responding==true) | select(.extensions.ip_address.data[]==(%v)) | .id' ", tkn, fqdn, ip)
+		hostid, _ := exec.RunWithStdout("bash", "-c", cmd)
+		hostid = strings.TrimSpace(strings.Trim(hostid, "\n"))
+		if len(hostid) == 0 {
+			zap.S().Infof("Unable to find host with IP %v please try again or run prep-node first", ip)
```
Comment (anupbarve) [inline cmd/attachNode.go:147]: Need to return from here with an error ?

File: cmd/attachNode.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	return hostIdsList, nil
+}
+
+func cluster_Status(exec cmdexec.Executor, fqdn string, token string, projectID string, clusterID string) string {
+	tkn := fmt.Sprintf(`"X-Auth-Token: %v"`, token)
+	cmd := fmt.Sprintf("curl -sH %v -X GET %v/qbert/v3/%v/clusters/%v | jq '.status' ", tkn, fqdn, projectID, clusterID)
+	status, _ := exec.RunWithStdout("bash", "-c", cmd)
+	status = strings.TrimSpace(strings.Trim(status, "\n\""))
+	return status
```
Comment (anupbarve) [inline cmd/attachNode.go:160]: What happens when this curl command fails ?
Need to check the status and handle error accordingly.

File: pkg/pmk/cluster.go [hunk 6]
```diff
@@ -53,12 +53,12 @@ func Bootstrap(ctx Config, c Client, req qbert.ClusterCreateRequest) error {
...
 	time.Sleep(ctx.WaitPeriod * time.Second)
-
+	var nodeIDs []string
+	nodeIDs = append(nodeIDs, nodeID)
 	zap.S().Info("Attaching node to the cluster...")
 	err = c.Qbert.AttachNode(
 		clusterID,
-		nodeID,
-		keystoneAuth.ProjectID, keystoneAuth.Token)
+		keystoneAuth.ProjectID, keystoneAuth.Token, nodeIDs, "workeer")
```
Comment (anupbarve) [inline pkg/pmk/cluster.go:61]: typo : workeer ?

**Review by anupbarve — CHANGES_REQUESTED** [review]
First pass.
Added some comments that need to be addressed.

File: cmd/attachNode.go (hunk 1 above)
Comment (devidask27) [reply cmd/attachNode.go]: Done

File: cmd/attachNode.go (hunk 2 above)
Comment (devidask27) [reply cmd/attachNode.go]: Added

File: cmd/attachNode.go (hunk 3 above)
Comment (devidask27) [reply cmd/attachNode.go]: Added

File: cmd/attachNode.go (hunk 4 above)
Comment (devidask27) [reply cmd/attachNode.go]: Done

File: cmd/attachNode.go (hunk 5 above)
Comment (devidask27) [reply cmd/attachNode.go]: Added error case

File: pkg/pmk/cluster.go (hunk 6 above)
Comment (devidask27) [reply pkg/pmk/cluster.go]: Fixed

**Review by anupbarve — CHANGES_REQUESTED** [review]
One thing that is missing in this implementation is that segment events are not added for both success or failure case.
Need to add it.

File: cmd/attachNode.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+			if err1 != nil {
+				if err := c.Segment.SendEvent("Attaching-node", auth, "Failed to attach worker node", ""); err != nil {
+					zap.S().Errorf("Unable to send Segment event for attach node. Error: %s", err.Error())
+				}
+				zap.S().Info("Encountered an error while attaching worker node to a Kubernetes cluster : ", err1)
+			} else {
+				if err := c.Segment.SendEvent("Attaching-node", auth, "Worker node attached", ""); err != nil {
+					zap.S().Errorf("Unable to send Segment event for attach node. Error: %s", err.Error())
+				}
+				zap.S().Info("Worker node(s) attached to cluster")
```
Comment (anupbarve) [inline cmd/attachNode.go:137]: Can you please add hostid to this log message ?

File: cmd/attachNode.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+			if err1 != nil {
+				if err := c.Segment.SendEvent("Attaching-node", auth, "Failed to attach master node", ""); err != nil {
+					zap.S().Errorf("Unable to send Segment event for attach node. Error: %s", err.Error())
+				}
+				zap.S().Info("Encountered an error while attaching master node to a Kubernetes cluster : ", err1)
+			} else {
+				if err := c.Segment.SendEvent("Attaching-node", auth, "Master node attached", ""); err != nil {
+					zap.S().Errorf("Unable to send Segment event for attach node. Error: %s", err.Error())
+				}
+				zap.S().Info("Master node(s) attached to cluster")
```
Comment (anupbarve) [inline cmd/attachNode.go:152]: Can you please add hostid to this log message ?

File: cmd/attachNode.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+func cluster_Status(exec cmdexec.Executor, fqdn string, token string, projectID string, clusterID string) string {
+	tkn := fmt.Sprintf(`"X-Auth-Token: %v"`, token)
+	cmd := fmt.Sprintf("curl -sH %v -X GET %v/qbert/v3/%v/clusters/%v | jq '.status' ", tkn, fqdn, projectID, clusterID)
+	status, err := exec.RunWithStdout("bash", "-c", cmd)
+	if err != nil {
+		zap.S().Fatalf("Unable to get cluster status : ", err)
+	}
+	status = strings.TrimSpace(strings.Trim(status, "\n\""))
```
Comment (anupbarve) [inline cmd/attachNode.go:185]: Log status here.

**Review by anupbarve — COMMENTED** [review]
Some minor comments. Else, looks good to me.
Please go ahead and add tests since the core logic is now reviewed.

**Review by anupbarve — APPROVED** [review]
Please update README.md before checkin.

**Pull request merged**